### ISSUE TITLE
Loop product and category thumbnails not correctly cropped

### ISSUE THREAD

**Issue by Jon007** [woocommerce/woocommerce#19961]
first reported on:
https://wordpress.org/support/topic/woo3-3-5-catalog-images-only-correct-size-in-customizer/
reposting here in order to submit pull request.

Tested with 3.3.5 and unmodified twenty-sixteen theme.

Since update from v3.1.2 to v3.3.5,
– catalogue pages no longer show thumbnails, they output reference to the full image, which may not be cropped causing visual issues
– if you open the customizer, the catalogue then shows the correct cropped images

not entirely sure if is bug in woocommerce or wordpress, but can be fixed safely from woocommerce end:

In wordpress core media.php the functions don’t work properly if passed a size string rather than array of size properties:
`function wp_get_attachment_image_src( $attachment_id, $size = 'thumbnail', $icon = false ) {`
calls
`function image_downsize( $id, $size = 'medium' ) {`
which expects to be passed array of size information, although it also accepts string, internally it calls:
`function image_get_intermediate_size( $post_id, $size = 'thumbnail' ) {`
When these functions are not passed an array of size information, if instead just passed a string such as ‘woocommerce_thumbnail’ it makes no effort to check for a match for a file of this size and simply returns false.
image_downsize then continues with the original size and calls:
`function image_constrain_size_for_editor( $width, $height, $size = 'medium', $context = null ) {`
which then returns a size which is constrained to the max size without cropping.

This can be easily fixed from the woocommerce end:

```
	function woocommerce_subcategory_thumbnail( $category ) {
		$small_thumbnail_size = apply_filters( 'subcategory_archive_thumbnail_size', 'woocommerce_thumbnail' );
		$dimensions           = wc_get_image_size( $small_thumbnail_size );
		$thumbnail_id         = get_woocommerce_term_meta( $category->term_id, 'thumbnail_id', true );

		if ( $thumbnail_id ) {
			$image        = wp_get_attachment_image_src( $thumbnail_id, $small_thumbnail_size );
```

needs to be changed to pass an array of width and height to all the 3 calls to the media.php functions:

```
	function woocommerce_subcategory_thumbnail( $category ) {
		$small_thumbnail_size = apply_filters( 'subcategory_archive_thumbnail_size', 'woocommerce_thumbnail' );
		$dimensions           = wc_get_image_size( $small_thumbnail_size );
        $wp_media_dimensions  = array($dimensions['width'],$dimensions['height']);
		$thumbnail_id         = get_woocommerce_term_meta( $category->term_id, 'thumbnail_id', true );

		if ( $thumbnail_id ) {
			$image        = wp_get_attachment_image_src( $thumbnail_id, $wp_media_dimensions );
			$image        = $image[0];
			$image_srcset = function_exists( 'wp_get_attachment_image_srcset' ) ? wp_get_attachment_image_srcset( $thumbnail_id, $wp_media_dimensions) : false;
			$image_sizes  = function_exists( 'wp_get_attachment_image_sizes' ) ? wp_get_attachment_image_sizes( $thumbnail_id, $wp_media_dimensions ) : false;
```

This resolves the problem for Category thumbnails.

to resolve the same problem for Products
 the product images in the loop are handled by 
```
	function woocommerce_template_loop_product_thumbnail() {
		echo woocommerce_get_product_thumbnail(); // WPCS: XSS ok.
	}
```

which calls `woocommerce_get_product_thumbnail()` which calls
`$product->get_image( $image_size )`
which is implemented in abstract-wc-product.php by a call to 
`get_the_post_thumbnail()`
which calls
`wp_get_attachment_image()`
all with the string size which generates the same problem as above for category images.

So the fix here is in abstract-wc-product.php 

```
	public function get_image( $size = 'woocommerce_thumbnail', $attr = array(), $placeholder = true ) {
		if ( has_post_thumbnail( $this->get_id() ) ) {
			$image = get_the_post_thumbnail( $this->get_id(), $size, $attr );
		} elseif ( ( $parent_id = wp_get_post_parent_id( $this->get_id() ) ) && has_post_thumbnail( $parent_id ) ) {
			$image = get_the_post_thumbnail( $parent_id, $wp_media_dimensions, $attr );
		} elseif ( $placeholder ) {
			$image = wc_placeholder_img( $wp_media_dimensions );
```

change to:

```
	public function get_image( $size = 'woocommerce_thumbnail', $attr = array(), $placeholder = true ) {
        $dimensions           = wc_get_image_size( $size );
        $wp_media_dimensions  = array($dimensions['width'],$dimensions['height']);
        if ( has_post_thumbnail( $this->get_id() ) ) {
			$image = get_the_post_thumbnail( $this->get_id(), $wp_media_dimensions, $attr );
		} elseif ( ( $parent_id = wp_get_post_parent_id( $this->get_id() ) ) && has_post_thumbnail( $parent_id ) ) {
			$image = get_the_post_thumbnail( $parent_id, $wp_media_dimensions, $attr );
		} elseif ( $placeholder ) {
			$image = wc_placeholder_img( $wp_media_dimensions );
```

tadaaa.. all fixed.

**Comment by claudiulodro** [comment]
Thanks for the thorough breakdown!

> function image_get_intermediate_size( $post_id, $size = 'thumbnail' ) {
> When these functions are not passed an array of size information, if instead just passed a string such as ‘woocommerce_thumbnail’ it makes no effort to check for a match for a file of this size and simply returns false.

I don't believe this is correct. You can see that [it checks for information in a defined size if possible](https://github.com/WordPress/WordPress/blob/3d4c461e504812689f97b7062f6da9d155448bc9/wp-includes/media.php#L749), so it should be picking up any generated thumbnails for that size. I'm not able to reproduce the issue you describe in my testing (Category thumbnails generate and use the 1:1 cropped thumbnail). Please let me know if I've missed something.

The frontend image thumbnail cropping happens through [a background process](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-regenerate-images.php#L36-L64), so there may be a short period where the uncropped images are being used while the cropped ones are getting generated. If they never generate it seems likely that the background processing is not happening for some reason, so that may be a good place to check.

**Comment by Jon007** [comment]
@claudiulodro thanks for the reply,

Worryingly I'm checking the size array in the wp_postmeta table and even for thumbnails of new products added since upgrade I'm only seeing the old sizes shop_catalog, shop_single, shop_thumbnail, not the new ones woocommerce_thumbnail, woocommerce_single, woocommerce_gallery_thumbnail, although by activating 'Simple Image Sizes' plugin I can see that these sizes are active.

And, what actually kicks off this regeneration process in a normal upgrade?
In this case no images actually need to be generated since files the right size already exist and no changes have been made to file size, it is simply that the woocommerce upgrade has added a new size name which isn't present in the existing data.

I do note in the init() of class-wc-regenerate-images there's a note and a condition
			// Regenerate thumbnails in the background after settings changes. **Not ran on multisite to avoid multiple simultanious jobs**.
It is a multisite (though only 1 site has woocommerce) and indeed since I haven't changed the size then the hash check in maybe_regenerate_images() would not trigger regeneration either.

I confess I'm also wary of the regeneration function - the site has many thousands more attachment image posts than products and I wasn't going to trigger a massive reprocessing without understanding why it should be necessary, when the image files already exist in the right size.

**Comment by claudiulodro** [comment]
@Jon007 good points. I'll take a look at your PR. It would definitely be best if the thumbnail regeneration used existing images of the correct size when possible.

**Comment by Jon007** [comment]
@claudiulodro I think the thumbnail regeneration will use the existing images if already available in the right size, it will just add extra entries pointing to them to the meta.

**Comment by mikejolley** [comment]
Passing image dimensions would work, we do it elsewhere, but it will interfere with our ability to target certain sizes, or take advantage of Jetpack photon and other image plugins. 

Interestingly we already have a fallback like your suggestion - this type of fallback logic runs regardless of whether it's multisite or not https://github.com/woocommerce/woocommerce/blob/dc7aa3069466f1b7e23fd7ef468760a6a9ddd241/includes/class-wc-regenerate-images.php#L92

That will return a similar sized image but only if it's larger than the current one. So say I had an image 150x150, and full size, and I wanted 250x250, it would return full. Thats how WordPress core handles it.

This works in the same way as your code which uses the image size array instead of a named size.

Trying to figure out what may cause that to fail for you.

**Comment by mikejolley** [comment]
Your fix does nothing btw. Even when you specify dimensions, it won't crop the full size image. It will only serve a larger, similar dimensioned image. If none exist, it won't do anything.

**Comment by Jon007** [comment]
Ah, the fallback is only executed if the size has changed:
`if ( ! self::image_size_matches_settings( $data, $size ) )`
The size has not changed so it does not run.

The size has not changed, it is a new size name added by woocommerce on upgrade which hadn't been added to the image metadata, although in fact an image of that size exists.

**Comment by mikejolley** [comment]
> Worryingly I'm checking the size array in the wp_postmeta table and even for thumbnails of new products added since upgrade I'm only seeing the old sizes shop_catalog, shop_single, shop_thumbnail, not the new ones woocommerce_thumbnail, woocommerce_single, woocommerce_gallery_thumbnail, although by activating 'Simple Image Sizes' plugin I can see that these sizes are active.

This may be related. The code I linked to in previous comment won't know the image size doesn't match settings if this meta is missing as we append a key called 'uncropped' to detect cropping mode changes. 

The new sizes are registered in the standard way. For those sizes to not generate I'd expect plugin/theme may be removing them or similar. I cannot explain it otherwise.

Notes:

- The provided patch does nothing different to core, we already have similar code in place. As such we don't need to merge it.
- I tested various options - it appears to be working to spec. I suspect missing sizes on your site are causing the glitch.

> if you open the customizer, the catalogue then shows the correct cropped images

There is some on-the-fly image creation there which actually creates the missing sizes. This isn't ran in the store however. Could be a bottleneck.
